extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/results/ExportResultsFrameProcessor.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.sql.http.ResultFormat;
+import org.apache.druid.storage.StorageConnector;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ExportResultsFrameProcessor implements FrameProcessor<Object>
+{
+  private final ReadableFrameChannel inputChannel;
+  private final ResultFormat exportFormat;
+  private final FrameReader frameReader;
+  private final StorageConnector storageConnector;
+  private final ObjectMapper jsonMapper;
+  private final int partitionNumber;
+  private final int workerNumber;
+  private final ChannelCounters channelCounter;
+
+  public ExportResultsFrameProcessor(
+      ReadableFrameChannel inputChannel,
+      ResultFormat exportFormat,
+      FrameReader frameReader,
+      StorageConnector storageConnector,
+      ObjectMapper jsonMapper,
+      int partitionNumber,
+      int workerNumber,
+      ChannelCounters channelCounter
+  )
+  {
+    this.inputChannel = inputChannel;
+    this.exportFormat = exportFormat;
+    this.frameReader = frameReader;
+    this.storageConnector = storageConnector;
+    this.jsonMapper = jsonMapper;
+    this.partitionNumber = partitionNumber;
+    this.workerNumber = workerNumber;
+    this.channelCounter = channelCounter;
+  }
+
+  @Override
+  public List<ReadableFrameChannel> inputChannels()
+  {
+    return Collections.singletonList(inputChannel);
+  }
+
+  @Override
+  public List<WritableFrameChannel> outputChannels()
+  {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public ReturnOrAwait<Object> runIncrementally(IntSet readableInputs) throws IOException
+  {
+    if (readableInputs.isEmpty()) {
+      return ReturnOrAwait.awaitAll(1);
+    }
+
+    if (inputChannel.isFinished()) {
+      return ReturnOrAwait.returnObject(Unit.instance());
+    } else {
+      exportFrame(inputChannel.read());
+      return ReturnOrAwait.awaitAll(1);
+    }
+  }
+
+  private void exportFrame(final Frame frame) throws IOException
+  {
+    final RowSignature signature = frameReader.signature();
+
+    final Sequence<Cursor> cursorSequence =
+        new FrameStorageAdapter(frame, frameReader, Intervals.ETERNITY)
+            .makeCursors(null, Intervals.ETERNITY, VirtualColumns.EMPTY, Granularities.ALL, false, null);
+
+    final String exportFilePath = getExportFilePath(workerNumber, partitionNumber, exportFormat);
+    try (OutputStream stream = storageConnector.write(exportFilePath)) {
+      ResultFormat.Writer formatter = exportFormat.createFormatter(stream, jsonMapper);
+
+      SequenceUtils.forEach(
+          cursorSequence,
+          cursor -> {
+            try {
+              formatter.writeResponseStart();
```

Comment:
Should this be written per cursor level or file level? If the former, then the current code is good, however, if it's the latter then this should be written outside the .forEach().
The FrameStorageAdapter returns a single cursor, so this doesn't make a difference, however, if that were to change in the future, the code should be correct.

Reply:
There's a `writeHeader()` field in the `formatter`'s class. Is that call not required? Perhaps the user would like the CSV to be populated with the row names in the header. Perhaps that can be a user-toggled input as well, something like what CSVParser does (hasHeaderRow). We can have a field like `writeHeaderRow`. This seems useful, particularly for CSV.

Reply:
Thanks for pointing this out! Moved the response start and end outside. Will look at adding a header with a separate PR.

Reply:
Cool! Would that require parser changes?